jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DelayedHandler.java

```diff
@@ -163,75 +205,152 @@ protected Callback getCallback()
                 Response.writeError(getRequest(), getResponse(), getCallback(), t);
             }
+            // return false to indicate the passed request/response/callback were not used.
+            return false;
         }
 
         protected abstract void delay() throws Exception;
     }
 
-    protected static class UntilContentDelayedProcess extends DelayedProcess
+    /**
+     * Delay dispatch until all content or an effective buffer size is reached
+     */
+    protected static class UntilContentDelayedProcess extends DelayedProcess implements Invocable.Task
     {
-        public UntilContentDelayedProcess(Handler handler, Request request, Response response, Callback callback)
+        private final Deque<Content.Chunk> _chunks = new ArrayDeque<>();
+        private final long _maxSize;
+        private long _estimatedSize;
+
+        /**
+         * @param handler The next handler
+         * @param request The delayed request
+         * @param response The delayed response
+         * @param callback The delayed callback
+         * @param maxSize The maximum size to buffer before dispatching to the next handler;
+         *                or -1 to use {@link HttpConnectionFactory#getInputBufferSize()}
+         */
+        public UntilContentDelayedProcess(Handler handler, Request request, Response response, Callback callback, long maxSize)
         {
             super(handler, request, response, callback);
+            _maxSize = maxSize < 0 ? request.getConnectionMetaData().getConnector().getConnectionFactory(HttpConnectionFactory.class).getInputBufferSize() : maxSize;
         }
 
         @Override
         protected void delay()
         {
-            Content.Chunk chunk = super.getRequest().read();
-            if (chunk == null)
-            {
-                getRequest().demand(org.eclipse.jetty.util.thread.Invocable.from(InvocationType.NON_BLOCKING, this::onContent));
-            }
-            else
+            read(false);
+        }
+
+        protected void read(boolean execute)
+        {
+            while (true)
             {
-                RewindChunkRequest request = new RewindChunkRequest(getRequest(), chunk);
-                try
+                Content.Chunk chunk = super.getRequest().read();
+                if (chunk == null)
+                {
+                    getRequest().demand(this);
+                    break;
+                }
+
+                // retain the chunk in the queue
+                if (!_chunks.add(chunk))
                 {
-                    getHandler().handle(request, getResponse(), getCallback());
+                    getCallback().failed(new IllegalStateException());
+                    break;
                 }
-                catch (Throwable x)
+
+                // Estimated size is 8 byte framing overhead per chunk plus the chunk size
+                _estimatedSize += 8 + chunk.remaining();
```

Comment:
I'd remove this attempt, it is wrong for every protocol and just complicates things to try to stick to the "input buffer" model, which is IMHO artificial.

Let's just count content bytes precisely.

Reply:
No.  Because  that will frequently result in very poor optimization.    If both the inputBufferSize and the maxSize are the same, then any framing overhead may result in overflowing the buffer and allocating a second mostly empty buffer.

Reply:
I've made it configurable, so if somebody wants an exact size, they can set it to 0.

Reply:
Not another configuration parameter!

I don't understand why are we using the input buffer?

The `DelayHandler` is about delaying until content.
Used to be first byte.
Why are we complicating things so much?

> If both the inputBufferSize and the maxSize are the same, then any framing overhead may result in overflowing the buffer and allocating a second mostly empty buffer.

That's exactly the problem of using the input buffer at this high level.
The framing will change for each protocol, and it will never be right, and making the framing configurable is not going to help.

At the `Handler` level we should not worry about framing of the underlying protocol.

Reply:
Not having an allowance for framing means that it will always be wrong as being an underestimate, which is bad. 

Having a rough approximation of 8 is probably an overestimate, which is not big problem. It is better to slightly under utilise a buffer rather than allocate another one.

As I've said several times, using the input buffer as a size is to allow zero configuration. The handler will have no additional memory commitment and thus the user does not need to think about it. Having a heuristic default means that it will often to be correct. You are suggesting something that has to mostly be configured, else it will mostly be wrong.

Reply:
@sbordet I wrote a test where the handler retained content until it had 8K of data, but it ignored any framing overhead.   The actual amount of buffers used were:
```
HTTP 49,152 > 8,192
HTTPS 61,440 > 8,192
H2C 81,920 > 8,192
H2 81,920 > 8,192
H3 24,576 > 8,192
```

Not allowing for framing overhead means that we can retain an order of magnitude more memory than is configured!

So if we do not account for framing overhead, a user is going to configure this handler with a "buffer size" of 8K and then be completely surprised that they are using up to 10 times that amount.   The reason is that this handler is not buffering, but retaining.  It is an entirely different commitment to make, and one that is difficult to explain to a user who just wants a casual optimization.     If the default configuration for this handler is -1 meaning 1 input buffer with allowance for framing, then they will have no additional memory commitment and there is nothing to explain other than we pick a good value for them.    If the default is some specific value without an allowance for framing, then we have to explain what that value means, and explain the framing overhead that they will incur and tell them that we give them no mechanism at all to deal with that framing overhead and they just have to be prepared for an order of magnitude more memory commitment than they configure in this handler.